src/fidesops/task/filter_element_match.py

```diff
@@ -0,0 +1,127 @@
+import copy
+import logging
+from collections import defaultdict
+
+from typing import List, Any, Dict
+
+import pydash
+
+from fidesops.graph.config import FieldPath
+from fidesops.task.refine_target_path import (
+    build_incoming_refined_target_paths,
+    DetailedPath,
+)
+
+logger = logging.getLogger(__name__)
+
+
+def filter_element_match(
+    row: Dict[str, Any], query_paths: Dict[FieldPath, List[Any]]
+) -> Dict[str, Any]:
+    """
+    Modifies row in place to only contain matched elements or matched embedded documents within arrays.
+
+    :param row: Record retrieved from a dataset
+    :param query_paths: FieldPaths mapped to query values
+    :return: A modified record with array elements potentially eliminated if array data was targeted by a query path
+
+    :Example:
+    The given row was retrieved from a dataset because an element in row["A"] matched 2 or an element in
+    row["C"]["D"] matched 5.  row["A"] is filtered to just contain the matching element, and row["C"] is filtered
+    to just contain the objects where "D" = 5.
+
+    filter_element_match(
+        row={"A": [1, 2, 3], "B": 2, "C": [{"D": 3, "E": 4}, {"D": 5, "E": 6}, {"D": 5, "E": 7}]},
+        query_paths={FieldPath("A"): [2], FieldPath("C, "D"): 5}
+    )
+
+    {"A": [2], "B": 2, "C": [{"D": 5, "E": 6}, {"D": 5, "E": 7}]}
+    """
+    detailed_target_paths: List[DetailedPath] = build_incoming_refined_target_paths(
+        row, query_paths
+    )
+
+    array_paths_to_preserve: Dict[str, List[int]] = _expand_array_paths_to_preserve(
+        detailed_target_paths
+    )
+
+    return _remove_paths_from_row(row, array_paths_to_preserve)
+
+
+def _remove_paths_from_row(
+    row: Dict[str, Any], preserve_indices: Dict[str, List[int]]
+) -> Dict[str, Any]:
+    """
+    Used by filter_element_match, remove array elements from row that are not specified in preserve_indices
+
+    :param row: Record retrieved from a dataset
+    :param preserve_indices: A dictionary of dot-separated paths to arrays, where the values are the list of indices
+    we want to *keep*
+    :return: A filtered row that has removed non-specified indices from arrays
+
+    :Example:
+    The first element in row["A"]["B"] was the only one specified to preserve, so we remove the other two.
+    _remove_paths_from_row(_remove_paths_from_row({"A": {"B": [{"C": "D"}, {"C": "F"}, {"C": "G"}]}}, {"A.B": [0]})
```

Comment:
Would be good to include a simpler example where the function is only called once, too

Reply:
oh sorry typo in the docstring, thank you